### DISCUSSION TITLE
OS and Linux distribution support requests

### DISCUSSION THREAD

**Discussion by Phundrak** [Phundrak/eshell-info-banner.el#16] [General]
While #2 standardized the retrieval of the name of the running Linux distribution, `eshell-info-banner.el` has yet to support non-standard Linux distributions (such as GUIX) and non-Linux OSes, such as macOS and Windows.
Here will be listed Linux distributions and OSes not yet supported:
- [X] macOS
  [Link to keep up with macOS versions](https://apple.fandom.com/wiki/List_of_Mac_OS_versions#Mac_OS_X_.2F_macOS)
  - [X] OS name and version 0f39c3188499ae1b2d0da96aa0bb42b771e18c66
  - [X] Kernel version 0f39c3188499ae1b2d0da96aa0bb42b771e18c66
  - [X] Partitions detection 3934c4e472b51bc0cc36811d4b2dc162ecbd1f81
  - [X] Memory display 46ada1b2649b1842534d8bf4c547c2648cbf2220
- [ ] Windows
  - [X] OS name and version #6
  - [ ] Partitions detection
    - [X] if [`duf`](https://github.com/muesli/duf) is installed on the system, d35516a3540d440441d64c9d2f6cf949ba22d6a3
    - [ ] native solution
  - [ ] Memory display
- [ ] BSDs
  - [X] OS name and version 0a8a340260dc9d21de135675dba49de1d1ed99a6
  - [X] Kernel version 0f39c3188499ae1b2d0da96aa0bb42b771e18c66
  - [ ] Partitions detection
    - [X] FreeBSD 763d459a9e5fb476017b91ec7ddfd3713d49a79c
    - [X] NetBSD a2f24e4633786cedbb396729c74db785b0fb7138
    - [X] DragonFly BSD 36b964a99311c95850ea14b6dc54f469d7b4e8b7
    - [ ] OpenBSD untested
  - [ ] Memory display
    - [X] FreeBSD 0a8a340260dc9d21de135675dba49de1d1ed99a6
    - [X] NetBSD 382bb550647c325321998c1411db07fbd862fc34
    - [X] DragonFly BSD 36b964a99311c95850ea14b6dc54f469d7b4e8b7
    - [ ] OpenBSD untested
  - [X] Fix uptime for NetBSD 89f78f023a8b2d6cf9b7e71d543063cdf23b7ea8
- [X] Hurd
  - [X] OS name and version 0f39c3188499ae1b2d0da96aa0bb42b771e18c66
  - [X] Kernel version 0f39c3188499ae1b2d0da96aa0bb42b771e18c66
  - [X] Partitions detection 763d459a9e5fb476017b91ec7ddfd3713d49a79c
  - [X] Memory display 763d459a9e5fb476017b91ec7ddfd3713d49a79c
- [ ] Linux distributions
  - [X] GUIX 0f39c3188499ae1b2d0da96aa0bb42b771e18c66
  - [X] Bedrock Linux c8073faa880791720a5b0550b7f502064542b464
  - [X] Proxmox c8073faa880791720a5b0550b7f502064542b464
  - [X] Android 0c9e6de3ee8657217c01468525acf19208e1a56d
  - [ ] ChromeOS
  - [ ] and others
- [ ] Haiku (Emacs 29 will support it)
  - [ ] OS Name and version
  - [ ] Kernel version
  - [ ] Partitions detection
  - [ ] Memory display

This list is subject to changes in the future, by shrinking or expanding.

**Comment by bkaestner** [comment]
>    * Windows:
>       
>       * ~OS name and version~ #6
>       * Partitions detection
>       * Memory display

You can use `wmic` for both partitions and memory:

```cmd
wmic logicaldisk get Caption,Size,Freespace,Name,Filesystem /FORMAT:csv

Node,Caption,FileSystem,FreeSpace,Name,Size
BensPC,C:,NTFS,50240790528,C:,249083572224
BensPC,D:,NTFS,509776490496,D:,3000457228288

wmic os get TotalVisibleMemorySize, FreePhysicalMemory /FORMAT:csv

Node,FreePhysicalMemory,TotalVisibleMemorySize
BensPC,10777156,16706896
```
Unfortunately, `wmic` is both deprecated and somewhat unreliable: most calls take <30ms, but some take more than 5 seconds. Maybe someone else with more Windows knowledge can provide a better alternative. Feel free to ping me if you're still interested in an initial version based on `wmic`.

**Comment by bkaestner** [comment]
> Unfortunately, `wmic` is both deprecated and somewhat unreliable

This might be my machine only, though. I have a network share mounted, but the server is offline. Windows still tries to reach it, check its properties,  and times out after some seconds. This might also impede other solutions on other operating systems, e.g. `nfs`, `sshfs` or similar shares. That being said, if you only want to check the "usual" mounts, we can probably use Emacs' internal tools:
```lisp
(defcustom eshell-info-banner-partition-windows-volumes
  (mapcar (lambda (x) (format "%c:" x)) (number-sequence ?C ?F))
  "Checked drive letters for volume measurements."
  :group 'eshell-info-banner
  :type '(repeat (string :tag "Volume letter")))

(defun windows-size-example ()
  "Returns a plist of the drives in `eshell-info-banner-partition-windows-volumes' with 
their respective sizes.

For every drive in the aforementioned variable, the resulting plist contains the 
following properties:

  label LABEL: the entry from `eshell-info-banner-partition-windows-volumes'
  free  BYTES: either the number of available bytes or nil
  size  BYTES: either the total disk size or nil on error

In order to format the bytes, you can use `file-size-human-readable-iec'."
  (mapcar (lambda (d)
            (let* ((info (file-system-info d))
                   (size (and info (nth 0 info)))
                   (free (and info (nth 1 info))))
              `(label ,d size ,size free ,free)))
          eshell-info-banner-partition-windows-volumes))

(mapconcat (lambda (x)
             (if (plist-get x 'size)
                 (format "Drive %s %s / %s\n"
                         (plist-get x 'label)
                         (file-size-human-readable-iec (plist-get x 'free))
                         (file-size-human-readable-iec (plist-get x 'size)))
               ""))             
           (windows-size-example)
           "")
```
The resulting string on the example above would be
```none
Drive C: 46.5 GiB / 232 GiB
Drive D: 474.8 GiB / 2.7 TiB
```
Again, feel free to ping me if I should prepare a PR based on the solution above or `wmic`.

**Comment by Phundrak** [comment]
Hello,

Apologies for the (very) late answer.

I’ll admit I’m a bit reluctant on using wmic since it is deprecated,
and as you pointed out it is not reliable. But I guess it could work
for some users, so something that could be done is another custom
variable like `eshell-info-banner-display-windows-partitions-p` so
that eshell-info-banner displays Windows partitions only when non-nil.
That way if it is too unreliable for some users, or even if it gets
removed entirely in future Windows updates, users can simply turn it
off and people for whom it works and/or who use older Windows versions
will still be able to use it.

**Comment by gopar** [comment]
Looks like Mac os has versions `11.5` and `11.6`. Seems like simple fixes to include in the list from the looks of it :)

**Comment by Phundrak** [comment]
It shold be, indeed. Are those still macOS Big Sur, or is it a new name?

--
Lucien “Phundrak” Cartier-Tilet

<https://phundrak.com> (Français)

<https://phundrak.com/en> (English)

Sent from GNU/Emacs

**Comment by gopar** [comment]
Still Big Sur :)
https://support.apple.com/en-us/HT211896

**Comment by Phundrak** [comment]
Alright, I’m updating the list. Thanks for your help!

--
Lucien “Phundrak” Cartier-Tilet

<https://phundrak.com> (Français)

<https://phundrak.com/en> (English)

Sent from GNU/Emacs
